Modules/Filtering/FFT/include/itkFFTImageFilterFactory.h

```diff
@@ -0,0 +1,139 @@
+  {
+    return "An FFTImageFilter factory";
+  }
+
+  /** Method for class instantiation. */
+  itkFactorylessNewMacro(Self);
+
+  /** Run-time type information (and related methods). */
+  itkTypeMacro(FFTImageFilterFactory, itk::ObjectFactoryBase);
+
+  /** Register one factory of this type  */
+  static void
+  RegisterOneFactory()
+  {
+    FFTImageFilterFactory::Pointer factory = FFTImageFilterFactory::New();
+
+    ObjectFactoryBase::RegisterFactoryInternal(factory);
+  }
+
+protected:
+  /** Override base FFT image filter constructor with the given TFFTImageFilter template
+   * with distinct specialization.
+   * Requires TFFTImageFilter to have the template signature <InputImageType, OutputImageType>.
+   */
+  template <typename InputPixelType, typename OutputPixelType, size_t ImageDimension>
+  void
+  OverrideFFTImageFilterType()
+  {
+    using InputImageType = Image<InputPixelType, ImageDimension>;
+    using OutputImageType = Image<OutputPixelType, ImageDimension>;
+    this->RegisterOverride(typeid(typename TFFTImageFilter<InputImageType, OutputImageType>::Superclass).name(),
+                           typeid(TFFTImageFilter<InputImageType, OutputImageType>).name(),
+                           "FFT Image Filter Override",
+                           true,
+                           CreateObjectFunction<TFFTImageFilter<InputImageType, OutputImageType>>::New());
+  }
+
+  FFTImageFilterFactory()
+  {
+    OverrideFFTImageFilterType<typename FFTImageFilterTraits<TFFTImageFilter>::template InputPixelType<float>,
+                               typename FFTImageFilterTraits<TFFTImageFilter>::template OutputPixelType<float>,
+                               1>();
+    OverrideFFTImageFilterType<typename FFTImageFilterTraits<TFFTImageFilter>::template InputPixelType<float>,
+                               typename FFTImageFilterTraits<TFFTImageFilter>::template OutputPixelType<float>,
+                               2>();
+    OverrideFFTImageFilterType<typename FFTImageFilterTraits<TFFTImageFilter>::template InputPixelType<float>,
+                               typename FFTImageFilterTraits<TFFTImageFilter>::template OutputPixelType<float>,
+                               3>();
+
+    OverrideFFTImageFilterType<typename FFTImageFilterTraits<TFFTImageFilter>::template InputPixelType<double>,
+                               typename FFTImageFilterTraits<TFFTImageFilter>::template OutputPixelType<double>,
+                               1>();
+    OverrideFFTImageFilterType<typename FFTImageFilterTraits<TFFTImageFilter>::template InputPixelType<double>,
+                               typename FFTImageFilterTraits<TFFTImageFilter>::template OutputPixelType<double>,
+                               2>();
+    OverrideFFTImageFilterType<typename FFTImageFilterTraits<TFFTImageFilter>::template InputPixelType<double>,
+                               typename FFTImageFilterTraits<TFFTImageFilter>::template OutputPixelType<double>,
```

Comment:
I am not sure about this thought, so I suggest putting it off until a later pull request.

Some of these FFTs are for 1D only.  Also, perhaps, others will be encouraged to be strictly more than 1D so that the user will use the 1D version in that case.  _If the proper thing to do in such a case_ is to not override the disfavored image dimensions then we can add a `static constexpr unsigned int MinImageDimension` and `static constexpr unsigned int MaxImageDimension` to `FFTImageFilterTraits`, so that `FFTImageFilterFactory()` knows exactly which ones to override.  (However, I am not so sure that that is the proper thing to do.  Even in that scenario perhaps these overrides are better than the defaults.)

Reply:
I added some of these changes to address compiler warnings, but others may have been solved by forward declaring the traits base class and may no longer be required. The second `template` in these lines is definitely required, I'll experiment with removing the `typename` keyword here.

In this case image dimension !== FFT dimension, we will want to allow 1D FFTs on ND images and vice versa. I don't think we should make changes that would couple these two ideas.

Reply:
To the extent that GCC 10.3.0 is a useful example ... the `template` keyword is not needed in the calls to `OverrideFFTImageFilterType`:

```c++
#include <complex>

template <typename T>
struct MyTraits
{};


template <>
struct MyTraits<int>
{
  template <typename TT>
  using MyType = std::complex<TT>;
};


int
main()
{
  MyTraits<int>::MyType<double>   i = 7; // works; i is of type double.
  MyTraits<float>::MyType<double> f = 7; // fails to compile: ‘MyType’ is not a member of ‘MyTraits<float>’
}
```

Reply:
Circling back around to this, with GCC 9.3.0 I get an error if `typename` is not included:
> /home/tom/repos/ITK/Modules/Filtering/FFT/include/itkFFTImageFilterFactory.h:123:80: note: say ‘typename itk::FFTImageFilterTraits<TFFTImageFilter>::InputPixelType<double>’ if a type is meant
/home/tom/repos/ITK/Modules/Filtering/FFT/include/itkFFTImageFilterFactory.h:126:80: error: dependent-name ‘itk::FFTImageFilterTraits<TFFTImageFilter>::InputPixelType<double>’ is parsed as a non-type, but instantiation yields a type
  126 |     OverrideFFTImageFilterType<FFTImageFilterTraits<TFFTImageFilter>::template InputPixelType<double>,
      |                                ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~
/home/tom/repos/ITK/Modules/Filtering/FFT/include/itkFFTImageFilterFactory.h:126:80: note: say ‘typename itk::FFTImageFilterTraits<TFFTImageFilter>::InputPixelType<double>’ if a type is meant
/home/tom/repos/ITK/Modules/Filtering/FFT/include/itkFFTImageFilterFactory.h: In instantiation of ‘itk::FFTImageFilterFactory<TFFTImageFilter>::FFTImageFilterFactory() [with TFFTImageFilter = itk::VnlRealToHalfHermitianForwardFFTImageFilter]’:
/home/tom/repos/ITK/Modules/Filtering/FFT/include/itkFFTImageFilterFactory.h:76:3:   required from ‘static itk::FFTImageFilterFactory<TFFTImageFilter>::Pointer itk::FFTImageFilterFactory<TFFTImageFilter>::New() [with TFFTImageFilter = itk::VnlRealToHalfHermitianForwardFFTImageFilter; itk::FFTImageFilterFactory<TFFTImageFilter>::Pointer = itk::SmartPointer<itk::FFTImageFilterFactory<itk::VnlRealToHalfHermitianForwardFFTImageFilter> >]’
/home/tom/repos/ITK/Modules/Core/Common/include/itkObjectFactoryBase.h:233:44:   required from ‘static void itk::ObjectFactoryBase::RegisterInternalFactoryOnce() [with TFactory = itk::FFTImageFilterFactory<itk::VnlRealToHalfHermitianForwardFFTImageFilter>]’
/home/tom/repos/ITK/Modules/Filtering/FFT/src/itkVnlFFTImageFilterFactories.cxx:82:118:   required from here
/home/tom/repos/ITK/Modules/Filtering/FFT/include/itkFFTImageFilterFactory.h:110:80: error: dependent-name ‘itk::FFTImageFilterTraits<TFFTImageFilter>::InputPixelType<float>’ is parsed as a non-type, but instantiation yields a type
  110 |     OverrideFFTImageFilterType<FFTImageFilterTraits<TFFTImageFilter>::template InputPixelType<float>,
      |                                ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~

[CTests on macOS](https://open.cdash.org/viewBuildError.php?buildid=7569604) have previously failed where the second `template` keyword is not included:
> Modules/Filtering/FFT/include/itkFFTImageFilterFactory.h:93:41: error: expected a qualified name after 'typename'

> Modules/Filtering/FFT/include/itkFFTImageFilterFactory.h:96:41: error: no template named 'FFTImageFilterTraits'     OverrideFFTImageFilterType<typename FFTImageFilterTraits<TFFTImageFilter>::InputPixelType<float>,                                         ^ Modules/Filtering/FFT/include/itkFFTImageFilterFactory.h:96:41: error: expected a qualified name after 'typename' Modules/Filtering/FFT/include/itkFFTImageFilterFactory.h:99:41: error: no template named 'FFTImageFilterTraits'     OverrideFFTImageFilterType<typename FFTImageFilterTraits<TFFTImageFilter>::InputPixelType<float>,                                         ^ Modules/Filtering/FFT/include/itkFFTImageFilterFactory.h:99:41: error: expected a qualified name after 'typename' Modules/Filtering/FFT/include/itkFFTImageFilterFactory.h:103:41: error: no template named 'FFTImageFilterTraits'     OverrideFFTImageFilterType<typename FFTImageFilterTraits<TFFTImageFilter>::InputPixelType<double>, The maximum number of reported warnings or errors has been reached!!!
--

It seems that the compiler will parse `InputPixelType` as a non-type if the `typename` keyword is not included and as a non-templated type if `typename` is included but `template` is not. Both are required here.

Reply:
I have learned something here ... thank you.